src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -1509,13 +1527,15 @@ static int client_parse_message(
         }
 
         r = dhcp_option_parse(message, len, dhcp_lease_parse_options, lease, &error_message);
-        if (r < 0)
+        if (r == -ENOMSG && client->bootp)
+                log_dhcp_client(client, "No DHCP options found in bootp message, continuing.");
+        else if (r < 0)
                 return log_dhcp_client_errno(client, r, "Failed to parse DHCP options, ignoring: %m");
```

Comment:
Parsing options for BOOTP reply sounds spurious. How about something like the following?
```c
if (client->bootp)
        r = DHCP_ACK;
else
        r = dhcp_option_parse(...);
if (r < 0)
        return log_dhcp_client_errno(...);
```

Reply:
> Parsing options for BOOTP reply sounds spurious.

Those replies can contain BOOTP Vendor Extensions https://www.rfc-editor.org/rfc/rfc2132
> All of the vendor information extensions defined in [RFC 1497](https://www.rfc-editor.org/rfc/rfc1497) [[2](https://www.rfc-editor.org/rfc/rfc2132#ref-2)] may
   be used as DHCP options.  The definitions given in [RFC 1497](https://www.rfc-editor.org/rfc/rfc1497) are
   included in this document, which supersedes [RFC 1497](https://www.rfc-editor.org/rfc/rfc1497).  All of the
   DHCP options defined in this document, except for those specific to
   DHCP as defined in [section 9](https://www.rfc-editor.org/rfc/rfc2132#section-9), may be used as BOOTP vendor information
   extensions.

Reply:
Thanks. So, BOOTP messages also supports some options. Hm, interesting.

Reply:
Yes, @evverx is correct. I need to run the parse options to successfully extract the address and server address. I added a comment in the updated code as to why I'm ignoring `-ENOMSG`

Reply:
> parse options to successfully extract the address and server address

I don't think it should process and apply DHCP-specific options though.
> All of the
   DHCP options defined in this document, except for those specific to
   DHCP as defined in [section 9](https://www.rfc-editor.org/rfc/rfc2132#section-9), may be used as BOOTP vendor information
   extensions.

I think the same applies to outgoing options. Right now with `Bootp=yes` the client sends the following options:
```sh
    Option: (55) Parameter Request List                                                                                                                       
        Length: 12                                                                                                                                            
        Parameter Request List Item: (1) Subnet Mask         
        Parameter Request List Item: (3) Router 
        Parameter Request List Item: (6) Domain Name Server                                                                                                   
        Parameter Request List Item: (12) Host Name
        Parameter Request List Item: (15) Domain Name
        Parameter Request List Item: (33) Static Route
        Parameter Request List Item: (42) Network Time Protocol Servers
        Parameter Request List Item: (101) TCode
        Parameter Request List Item: (114) DHCP Captive-Portal
        Parameter Request List Item: (120) SIP Servers
        Parameter Request List Item: (121) Classless Static Route
        Parameter Request List Item: (162) Unassigned
    Option: (57) Maximum DHCP Message Size
        Length: 2
        Maximum DHCP Message Size: 1472 
    Option: (80) Rapid commit
        Length: 0
        Value: <MISSING>
    Option: (12) Host Name
        Length: 4
        Host Name: YOLO
    Option: (255) End
        Option End: 255
```
All those options except for the Host Name Option (and the End option) are DHCP-specific.

It also complains about plain BOOTP replies messages somewhere:
```sh
DHCPv4 client: received lease lacks address, server address or lease lifetime, ignoring.
```

I can't say I read all the RFC but it doesn't look like the client follows them. I still think it's kind of confusing that the BOOTP client is implemented in the DHCP section because I'd interpret it as a DHCP client that can understand BOOTP replies (but it's a BOOTP client).